File: src/main/resources/php/PasswordUtils.php
<?php
/**
 * Utility class for password hashing and verification.
 * Uses SHA-256 with salt for secure password storage.
 * This implementation is compatible with the Java equivalent.
 */
class PasswordUtils {
    
    // The length of the salt in bytes
    private const SALT_LENGTH = 16;
    
    // The algorithm to use for hashing
    private const ALGORITHM = 'sha256';
    
    // The number of iterations for the hashing
    private const ITERATIONS = 10000;
    
    /**
     * Hashes a password using SHA-256 with salt.
     * 
     * @param string $password The password to hash
     * @return string The hashed password with salt in format "salt:hash"
     */
    public static function hashPassword($password) {
        if (empty($password)) {
            throw new \InvalidArgumentException("Password cannot be null or empty");
        }
        
        // Generate a random salt
        $salt = random_bytes(self::SALT_LENGTH);
        
        // Hash the password with the salt
        $hash = self::hashWithSalt($password, $salt);
        
        // Convert salt and hash to Base64 for storage
        $saltStr = base64_encode($salt);
        $hashStr = base64_encode($hash);
        
        // Return the salt and hash combined, separated by a colon
        return $saltStr . ':' . $hashStr;
    }
    
    /**
     * Verifies that a password matches a hashed password.
     * 
     * @param string $password The password to check
     * @param string $hashedPassword The hashed password to check against (in format "salt:hash")
     * @return bool true if the password matches, false otherwise
     */
    public static function verifyPassword($password, $hashedPassword) {
        if (empty($password) || empty($hashedPassword)) {
            return false;
        }
        
        try {
            // Split the stored string to get the salt and hash
            $parts = explode(':', $hashedPassword);
            if (count($parts) !== 2) {
                return false;
            }
            
            $salt = base64_decode($parts[0]);
            $storedHash = base64_decode($parts[1]);
            
            if ($salt === false || $storedHash === false) {
                return false;
            }
            
            // Hash the input password with the same salt
            $hash = self::hashWithSalt($password, $salt);
            
            // Compare the calculated hash with the stored hash
            // Use hash_equals for constant-time comparison to prevent timing attacks
            return hash_equals($storedHash, $hash);
        } catch (\Exception $e) {
            // If there's any error in decoding or hashing
            return false;
        }
    }
    
    /**
     * Internal method to hash a password with a given salt.
     * 
     * @param string $password The password to hash
     * @param string $salt The salt to use
     * @return string The hashed password
     */
    private static function hashWithSalt($password, $salt) {
        // Perform the initial hash
        $hash = hash(self::ALGORITHM, $salt . $password, true);
        
        // Perform multiple iterations to increase security
        for ($i = 0; $i < self::ITERATIONS - 1; $i++) {
            $hash = hash(self::ALGORITHM, $hash, true);
        }
        
        return $hash;
    }
    
    /**
     * Checks if a password needs to be rehashed.
     * Always returns false in this implementation since we don't have version markers.
     * 
     * @param string $hashedPassword The hashed password to check
     * @return bool true if the password should be rehashed
     */
    public static function needsRehash($hashedPassword) {
        // This implementation doesn't support rehashing based on algorithm changes
        // since we don't store version information in the hash
        return false;
    }
}
?> 

File: src/main/java/utils/PasswordUtils.java
package utils;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Utility class for password hashing and verification.
 * Uses SHA-256 with salt for secure password storage.
 * This implementation is compatible with the PHP equivalent.
 */
public class PasswordUtils {
    
    // The length of the salt in bytes
    private static final int SALT_LENGTH = 16;
    
    // The algorithm to use for hashing
    private static final String ALGORITHM = "SHA-256";
    
    // The number of iterations for PBKDF2
    private static final int ITERATIONS = 10000;
    
    /**
     * Hashes a password using SHA-256 with salt.
     * 
     * @param password The password to hash
     * @return The hashed password with salt in format "salt:hash"
     */
    public static String hashPassword(String password) {
        if (password == null || password.isEmpty()) {
            throw new IllegalArgumentException("Password cannot be null or empty");
        }
        
        try {
            // Generate a random salt
            SecureRandom random = new SecureRandom();
            byte[] salt = new byte[SALT_LENGTH];
            random.nextBytes(salt);
            
            // Hash the password with the salt
            byte[] hash = hashWithSalt(password, salt);
            
            // Convert salt and hash to Base64 for storage
            String saltStr = Base64.getEncoder().encodeToString(salt);
            String hashStr = Base64.getEncoder().encodeToString(hash);
            
            // Return the salt and hash combined, separated by a colon
            return saltStr + ":" + hashStr;
        } catch (NoSuchAlgorithmException e) {
            // This should never happen as SHA-256 is a standard algorithm
            throw new RuntimeException("Error hashing password", e);
        }
    }
    
    /**
     * Verifies that a password matches a hashed password.
     * 
     * @param password The password to check
     * @param hashedPassword The hashed password to check against (in format "salt:hash")
     * @return true if the password matches, false otherwise
     */
    public static boolean verifyPassword(String password, String hashedPassword) {
        if (password == null || password.isEmpty() || hashedPassword == null || hashedPassword.isEmpty()) {
            return false;
        }
        
        try {
            // Split the stored string to get the salt and hash
            String[] parts = hashedPassword.split(":");
            if (parts.length != 2) {
                return false;
            }
            
            byte[] salt = Base64.getDecoder().decode(parts[0]);
            byte[] storedHash = Base64.getDecoder().decode(parts[1]);
            
            // Hash the input password with the same salt
            byte[] hash = hashWithSalt(password, salt);
            
            // Compare the calculated hash with the stored hash
            // This is a constant-time comparison to prevent timing attacks
            return MessageDigest.isEqual(hash, storedHash);
        } catch (IllegalArgumentException | NoSuchAlgorithmException e) {
            // If there's any error in decoding or hashing
            return false;
        }
    }
    
    /**
     * Internal method to hash a password with a given salt.
     * 
     * @param password The password to hash
     * @param salt The salt to use
     * @return The hashed password
     * @throws NoSuchAlgorithmException If the algorithm is not available
     */
    private static byte[] hashWithSalt(String password, byte[] salt) throws NoSuchAlgorithmException {
        // Create a message digest for SHA-256
        MessageDigest md = MessageDigest.getInstance(ALGORITHM);
        
        // Add the salt to the digest
        md.update(salt);
        
        // Add the password bytes to the digest
        byte[] passwordBytes = password.getBytes(StandardCharsets.UTF_8);
        
        // Perform multiple iterations to increase security
        byte[] hash = md.digest(passwordBytes);
        for (int i = 0; i < ITERATIONS - 1; i++) {
            md.reset();
            hash = md.digest(hash);
        }
        
        return hash;
    }
    
    /**
     * Checks if a password needs to be rehashed.
     * Always returns false in this implementation since we don't have version markers.
     * 
     * @param hashedPassword The hashed password to check
     * @return true if the password should be rehashed
     */
    public static boolean needsRehash(String hashedPassword) {
        // This implementation doesn't support rehashing based on algorithm changes
        // since we don't store version information in the hash
        return false;
    }
} 